### PR TITLE
Tx speed up

### PR DESCRIPTION
**Pull request by zhongqiuwood**
Closes: #XXX

## Description

______

For contributor use:

- [ ] Wrote tests
- [ ] Updated CHANGELOG_PENDING.md
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: libs/clist/clist.go [hunk 1]
```diff
@@ -405,3 +442,163 @@ func waitGroup1() (wg *sync.WaitGroup) {
...
+	if l.head == nil || l.tail == nil {
+		l.mtx.Unlock()
+		panic("Remove(e) on empty CList")
+	}
+	if prev == nil && l.head != ele {
+		l.mtx.Unlock()
+		panic("Remove(e) with false head")
+	}
+	if next == nil && l.tail != ele {
+		l.mtx.Unlock()
```
Comment (KamiD) [inline libs/clist/clist.go:553]: should delete if use ```defer l.mtx.Unlock```

File: mempool/clist_mempool.go [hunk 2]
```diff
@@ -67,6 +70,9 @@ type CListMempool struct {
 	logger log.Logger
 
 	metrics *Metrics
+
+	AddressRecord map[string]map[string]*clist.CElement // Address -> (txHash -> *CElement)
```
Comment (ilovers) [inline mempool/clist_mempool.go:74]:
```suggestion
	addressMap map[string]map[string]*clist.CElement // Address -> (txHash -> *CElement)
```

File: mempool/clist_mempool.go [hunk 3]
```diff
@@ -67,6 +70,9 @@ type CListMempool struct {
...
+	addrMapRWLock sync.RWMutex
```
Comment (ilovers) [inline mempool/clist_mempool.go:75]:
```suggestion
	addrMapMtx sync.RWMutex
```

File: libs/clist/clist.go [hunk 4]
```diff
@@ -405,3 +442,163 @@ func waitGroup1() (wg *sync.WaitGroup) {
...
 }
+
+// ===============================================
+
+// head -> tail: gasPrice(big -> samll)
+func (l *CList) InsertElement(ele *CElement) *CElement {
+	l.mtx.Lock()
+	defer func() {
+		l.mtx.Unlock()
+	}()
```
Comment (ilovers) [inline libs/clist/clist.go:453]:
```suggestion
	defer l.mtx.Unlock()
```

File: libs/clist/clist.go [hunk 5]
```diff
@@ -211,6 +216,37 @@ func (e *CElement) SetRemoved() {
...
+	// This wakes up anyone waiting in either direction.
+	if e.prev == nil {
+		e.prevWg.Done()
+		close(e.prevWaitCh)
+	}
+	if e.next == nil {
+		e.nextWg.Done()
+		close(e.nextWaitCh)
+	}
+	e.mtx.Unlock()
```
[the remark this answers is not in the document]
Comment (louisliu2048) [reply libs/clist/clist.go:231]: ignore, follow raw code style

File: libs/clist/clist.go [hunk 6]
```diff
@@ -211,6 +216,37 @@ func (e *CElement) SetRemoved() {
...
+}
+
+func (e *CElement) NewDetachPrev() {
+	e.mtx.Lock()
+	e.prev = nil
+	e.prevWg = waitGroup1()
+	e.prevWaitCh = make(chan struct{})
+	e.mtx.Unlock()
```
[the remark this answers is not in the document]
Comment (louisliu2048) [reply libs/clist/clist.go:239]: ignore, follow raw code style

File: libs/clist/clist.go [hunk 7]
```diff
@@ -211,6 +216,37 @@ func (e *CElement) SetRemoved() {
...
+}
+
+func (e *CElement) NewDetachNext() {
+	e.mtx.Lock()
+	e.next = nil
+	e.nextWg = waitGroup1()
+	e.nextWaitCh = make(chan struct{})
+	e.mtx.Unlock()
```
[the remark this answers is not in the document]
Comment (louisliu2048) [reply libs/clist/clist.go:247]: ignore, follow raw code style

File: libs/clist/clist.go (hunk 1 above)
Comment (louisliu2048) [reply libs/clist/clist.go:551]: ignore, follow raw code style

**Review by xiangjianmeng — APPROVED** [review]

File: mempool/clist_mempool.go (hunk 2 above)
Comment (louisliu2048) [reply mempool/clist_mempool.go]: good idea

File: mempool/clist_mempool.go (hunk 3 above)
Comment (louisliu2048) [reply mempool/clist_mempool.go]: good idea

File: libs/clist/clist.go (hunk 4 above)
Comment (louisliu2048) [reply libs/clist/clist.go]: yes, forget to move it out.

**Review by KamiD — APPROVED** [review]

**Review by ilovers — APPROVED** [review]

**Pull request merged**